### ISSUE TITLE
How to force git deployment after previous failure

### ISSUE THREAD

**Issue by dprothero** [projectkudu/kudu#1868]
There are times when the git push to kudu succeeds, but the deployment fails for some reason (e.g. #878). When this happens, and you try to do push again, you simply get "Everything up-to-date" and no deployment.

Other than making another commit to the repo and re-pushing, is there some way to trigger the deployment?

**Comment by davidebbo** [comment]
Indeed, that is an annoying situation, and I don't believe we have a great way of getting out of it today, short of a dummy push.

In theory, this only happens in very abnormal situations, like the weird Cygwin error.

**Comment by dprothero** [comment]
I don't know how abnormal the situations are... One instance was when I was troubleshooting private nuget packages. I was using the nuget command from the console to add a custom nuget package repository (a local directory)... if I didn't set that up right, the deploy would fail (not finding the nuget packages)... 

Another instance is when you are experimenting with msbuild settings using the AppSettings.

Basically, any time when it's an environmental issue that could have an affect on the success or failure of the deployment.

It would be nice if there was some API call or something we could make to force a redeployment from the current contents of the git repo. Feature request?

**Comment by davidebbo** [comment]
That does not sound right, and we should look more into the details of those failures. In theory, if the deployment logic is able to start in any way, there should be a failed deployment showing up in Portal that can be retried. But in rare cases like Cygwin error, it doesn't happen because it fails so early.

Are you sure that you see no deployment item that you could retry in your scenarios?

**Comment by sitereactor** [comment]
@dprothero isn't the `/api/deployments` endpoint what you are looking for (ref. https://github.com/projectkudu/kudu/wiki/REST-API#deployment)?

If you send an empty PUT request to the above endpoint it will deploy the HEAD of the repository: `PUT /api/deployments {}`

**Comment by dprothero** [comment]
That should do the trick, thanks!
